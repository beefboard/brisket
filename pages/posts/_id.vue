<template>
  <div class="page-container">
    <div class="page">
      <div class="top">
        <nuxt-link class="back" to="/">
          <fa :icon="faChevronLeft"/>Back to posts
        </nuxt-link>
        <div v-if="admin" class="admin">
          <button class="approve-button beefbutton" v-if="!post.approved" @click="approve">Approve</button>
          <button class="beefbutton" v-if="post.approved && !post.pinned" @click="pin">Pin</button>
          <button class="beefbutton" v-if="post.approved && post.pinned" @click="unpin">Unpin</button>
          <button class="beefbutton" @click="confirmRemove"><fa :icon="faTrashAlt"/></button>
        </div>
      </div>
      <post-details
        :post="post"
      />
    </div>
    <v-dialog/>
  </div>
</template>
<script>
import config from '~/nuxt.config'
import PostDetails from '~/components/PostDetails'
import { faChevronLeft, faTrashAlt } from '@fortawesome/free-solid-svg-icons'

export default {
  components: {
    PostDetails
  },
  async asyncData({ store, params, error, router }) {
    try {
      const post = await store.dispatch('getPost', params.id)
      const admin = store.state.auth && store.state.auth.admin

      return {
        id: params.id,
        api: config.axios.baseURL,
        admin: admin,
        post: post
      }
    } catch (e) {
      console.log(e)
      error({ statusCode: 404, message: 'Post not found' })
    }
  },
  validate({ redirect, params, $axios }) {
    if (!params.id) {
      return redirect('/')
    }
    return true
  },
  methods: {
    async approve() {
      try {
        await this.$store.dispatch('approvePost', this.id)
        this.post.approved = true
      } catch (e) {}
    },
    async pin() {
      try {
        await this.$store.dispatch('pinPost', {
          id: this.id,
          pinned: true
        })
        this.post.pinned = true
      } catch (_) {}
    },
    async unpin() {
      try {
        await this.$store.dispatch('pinPost', {
          id: this.id,
          pinned: false
        })
        this.post.pinned = false
      } catch (_) {}
    },
    async remove() {
      try {
        await this.$store.dispatch('deletePost', this.id)
        this.$router.push('/')
      } catch (_) {}
    },
    async confirmRemove() {
      this.$modal.show('dialog', {
        title: 'Delete post',
        text: 'Are you sure you want to delete this post?',
        buttons: [
          {
            title: 'Confirm',
            handler: this.remove
          },
          {
            title: 'Cancel'
          }
        ]
      })
    }
  },
  computed: {
    faChevronLeft() {
      return faChevronLeft
    },
    faTrashAlt() {
      return faTrashAlt
    }
  }
}
</script>

<style scoped>
.page {
  display: flex;
  flex-direction: column;
  flex: 1;
}

.top {
  margin: 1rem;
  display: flex;
  justify-content: space-between;
}

a {
  display: flex;
  align-items: center;
  color: black;
  text-decoration: none;
}

button {
  font-size: 1rem;
}
</style>
