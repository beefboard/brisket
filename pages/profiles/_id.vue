<template>
  <div class="container">
    <div class="profile-container">
      <div class="username">
        {{ username }}
      </div>
      <div class="name">{{ firstName }} {{ lastName }}</div>
    </div>
  </div>
</template>
<script>
export default {
  async asyncData({ params, $axios, store, error }) {
    if (params.id == store.state.auth.username) {
      return store.state.auth
    }

    try {
      return (await $axios.get(`/accounts/${params.id}`)).data
    } catch (e) {
      error({ statusCode: 404, message: 'User not found' })
    }
  },
  validate({ redirect, params }) {
    if (!params.id) {
      return redirect('/')
    }

    return true
  }
}
</script>

<style>
.container {
  margin: 1rem;
  display: flex;
  align-content: center;
  align-items: center;
  text-align: start;
  flex-direction: column;
}

.username {
  font-size: 3rem;
}

.name {
  font-size: 1rem;
}
</style>
