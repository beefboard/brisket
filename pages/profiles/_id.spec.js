import VueTestUtils from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import _id from './_id.vue'

describe('profiles/_id', () => {
  let mockUser
  let mockId

  beforeEach(() => {
    mockUser = {
      username: 'test',
      firstName: 'lel',
      lastName: 'lol'
    }
    mockId = 'asfdasf'
  })

  function renderLayout() {
    return shallowMount(_id, {
      data() {
        return mockUser
      },
      mocks: {
        params: {
          id: mockId
        }
      }
    })
  }
  it('mounts properly', () => {
    const wrapper = renderLayout()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('renders properly', () => {
    const wrapper = renderLayout()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('displays user details in page', () => {
    mockUser = {
      username: 'auser',
      firstName: 'aname',
      lastName: 'alastname'
    }

    const wrapper = renderLayout()
    expect(wrapper.find('.username').text()).toBe('auser')
    expect(wrapper.find('.name').text()).toBe('aname alastname')
  })

  it('Sets the page title to profile name', () => {
    const wrapper = renderLayout()

    // Mock this.username in head()
    wrapper.vm.$options.username = 'aname'

    const headData = wrapper.vm.$options.head()
    expect(headData.title).toBe("aname's profile - Beefboard")
  })

  it('navigates to home if no username in address', () => {
    const redirect = jest.fn()
    const mockParams = {}

    const wrapper = renderLayout()

    wrapper.vm.$options.validate({ redirect, params: mockParams })
    expect(redirect).toHaveBeenCalledWith('/')
  })

  it('does not navigate home if id in address', () => {
    const redirect = jest.fn()
    const mockParams = { id: 'asdasd' }

    const wrapper = renderLayout()

    redirect.mockReset()
    wrapper.vm.$options.validate({ redirect, params: mockParams })
    expect(redirect).not.toHaveBeenCalled()
  })

  test('asyncData uses current auth data if profile is our own', async () => {
    const wrapper = renderLayout()

    const error = jest.fn()
    const params = { id: 'testusername' }
    const mockStore = {
      state: {
        auth: {
          username: 'testusername',
          firstName: 'expectedFirstname'
        }
      },
      dispatch: jest.fn()
    }

    const userDetails = await wrapper.vm.$options.asyncData({
      store: mockStore,
      params,
      error
    })

    expect(userDetails.firstName).toBe('expectedFirstname')
  })

  test('asyncData gets user data from store', async () => {
    const wrapper = renderLayout()

    const error = jest.fn()
    const params = { id: 'testusername' }
    const mockStore = {
      state: {},
      dispatch: jest.fn()
    }

    await wrapper.vm.$options.asyncData({ store: mockStore, params, error })
    expect(mockStore.dispatch).toHaveBeenCalledWith('getUser', 'testusername')
  })

  test('asyncData calls 404 error if no user details found for id', async () => {
    const wrapper = renderLayout()

    const error = jest.fn()
    const params = { id: 'testusername' }
    const mockStore = {
      state: {},
      dispatch: async () => {
        throw {
          response: {
            status: 404
          }
        }
      }
    }

    await wrapper.vm.$options.asyncData({ store: mockStore, params, error })
    expect(error).toHaveBeenCalledWith({
      statusCode: 404,
      message: 'User not found'
    })
  })

  test('asyncData calls 500 error if api fails', async () => {
    const wrapper = renderLayout()

    const error = jest.fn()
    const params = { id: 'testusername' }
    const mockStore = {
      state: {},
      dispatch: async () => {
        throw new Error('any error')
      }
    }

    await wrapper.vm.$options.asyncData({ store: mockStore, params, error })
    expect(error).toHaveBeenCalledWith({
      statusCode: 500,
      message: 'Unknown error'
    })
  })
})
