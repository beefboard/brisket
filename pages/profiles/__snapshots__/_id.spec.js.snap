// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profiles/_id renders properly 1`] = `
<div class="page-container">
  <div class="profile-container">
    <div>
      <div class="username">
        test
      </div>
      <div class="name">lel lol</div>
    </div>
  </div>
</div>
`;
